gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageClientReadChannel.java

```diff
@@ -0,0 +1,505 @@
+      // No seek is required
+      return;
+    }
+
+    logger.atFiner().log(
+        "Performing lazySeek from %s to %s position for '%s'",
+        contentChannelPosition, currentPosition, resourceId);
+    // record the variation in channelPosition
+    // used to auto-detect random access
+    long oldPosition = contentChannelPosition;
+    long seekDistance = currentPosition - contentChannelPosition;
+    // Check if inPlace seek is a viable option
+    // If not close existing channel and open new one.
+
+    if (contentReadChannel != null
+        && seekDistance > 0
+        && seekDistance <= readOptions.getInplaceSeekLimit()
+        && currentPosition < contentChannelEnd) {
+      skipInPlace(seekDistance);
+    } else {
+      closeContentChannel();
+    }
+    if (contentReadChannel == null) {
+      if (isRandomAccessPattern(oldPosition)) {
+        setRandomAccess();
+      }
+      contentReadChannel = openReadChannel(bytesToRead);
+    }
+  }
+
+  // Update this function to create the reader and then set limits
+  private ReadableByteChannel getStorageReadChannel(long seek, long limit) throws IOException {
+    ReadChannel readChannel = storage.reader(blobId, generateReadOptions(blobId));
+    try {
+      readChannel.seek(seek);
+      // TODO: should we also add a check to not set limit if contentChannelEnd == objectSize?
+      if (limit != -1) {
+        readChannel.limit(limit);
+      }
+      return readChannel;
+    } catch (Exception e) {
+      throw new IOException(
+          String.format(
+              "Unable to update the boundaries/Range of contentChannel %s", resourceId.toString()),
+          e);
+    }
+  }
+
+  private static BlobSourceOption[] generateReadOptions(BlobId blobId) {
+    List<BlobSourceOption> readOptions = new ArrayList<>();
+    if (blobId.getGeneration() != null) {
+      readOptions.add(BlobSourceOption.generationMatch(blobId.getGeneration()));
+    }
+    return readOptions.toArray(new BlobSourceOption[readOptions.size()]);
+  }
+
+  private boolean isFooterRead() {
+    return objectSize - currentPosition <= footerSize;
+  }
+
+  private static void validate(GoogleCloudStorageItemInfo itemInfo) throws IOException {
+    checkNotNull(itemInfo, "itemInfo cannot be null");
+    StorageResourceId resourceId = itemInfo.getResourceId();
+    checkArgument(
+        resourceId.isStorageObject(), "Can not open a non-file object for read: %s", resourceId);
+    if (!itemInfo.exists()) {
+      throw new FileNotFoundException(String.format("File not found: %s", resourceId));
+    }
+    // The non-gRPC read channel has special support for gzip. This channel doesn't
+    // decompress gzip-encoded objects on the fly, so best to fail fast rather than return
+    // gibberish unexpectedly.
+    String contentEncoding = itemInfo.getContentEncoding();
+    if (contentEncoding != null && contentEncoding.contains("gzip")) {
```

Comment:
contentEncoding.contains("gzip") seem  like a weak check.

Reply:
Took it from existing check we have in `GrpcReadChannel` implementation. Any recommendation?

Reply:
why not equals?

Reply:
No header can have multiple `,` separated values.

Reply:
how can a file encoded in different ways? Can you give an example where there are multiple?

Reply:
https://www.holisticseo.digital/technical-seo/web-accessibility/http-header/content-encoding

Also, it's in alignment with what we already do [GoogleCloudStorageGrpcReadChannel].(https://github.com/GoogleCloudDataproc/hadoop-connectors/blob/701036eccddda54e0ef0b1e7d3514abb61dd87c9/gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcReadChannel.java#L199)